File: Android/wallet/src/main/java/com/flow/wallet/keys/SeedPhraseKey.kt
package com.flow.wallet.keys

import android.util.Log
import com.flow.wallet.NativeLibraryManager
import com.google.common.io.BaseEncoding
import com.flow.wallet.crypto.BIP39
import com.flow.wallet.crypto.ChaChaPolyCipher
import com.flow.wallet.crypto.HasherImpl
import com.flow.wallet.errors.WalletError
import com.flow.wallet.storage.StorageProtocol
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.onflow.flow.models.HashingAlgorithm
import org.onflow.flow.models.SigningAlgorithm
import wallet.core.jni.CoinType
import wallet.core.jni.Curve
import wallet.core.jni.HDWallet
import java.security.KeyFactory
import java.security.KeyPair
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import wallet.core.jni.PrivateKey as TWPrivateKey

@Serializable
private data class KeyData(
    @SerialName("mnemonic")
    val mnemonic: String,
    @SerialName("passphrase")
    val passphrase: String,
    @SerialName("path")
    val path: String,
    @SerialName("length")
    val length: BIP39.SeedPhraseLength
)

/**
 * Concrete implementation of SeedPhraseKeyProvider
 * Manages BIP39 seed phrase based keys using Trust WalletCore
 */
class SeedPhraseKey(
    private val mnemonicString: String,
    private val passphrase: String,
    override val derivationPath: String,
    private var keyPair: KeyPair?,
    override var storage: StorageProtocol,
    private val seedPhraseLength: BIP39.SeedPhraseLength = BIP39.SeedPhraseLength.TWELVE
) : SeedPhraseKeyProvider, EthereumKeyProtocol {
    companion object {
        private const val TAG = "SeedPhraseKey"
        private const val DEFAULT_DERIVATION_PATH = "m/44'/539'/0'/0/0"
        private const val ETH_DERIVATION_PREFIX = "m/44'/60'/0'/0"

        fun parseDerivationPath(path: String): IntArray {
            return path.split("/")
                .filter { it.isNotBlank() && it != "m" }
                .map {
                    when {
                        it.endsWith("'") -> it.dropLast(1).toInt() or 0x80000000.toInt()
                        else -> it.toInt()
                    }
                }
                .toIntArray()
        }
    }

    private val hdWallet: HDWallet = try {
        // Ensure native library is loaded before creating HD wallet
        if (!NativeLibraryManager.ensureLibraryLoaded()) {
            throw WalletError.InitHDWalletFailed
        }
        HDWallet(mnemonicString, passphrase)
    } catch (e: Exception) {
        Log.e(TAG, "Failed to initialize HD wallet", e)
        throw WalletError.InitHDWalletFailed
    }

    private fun getCurveForAlgorithm(signAlgo: SigningAlgorithm): Curve {
        return when (signAlgo) {
            SigningAlgorithm.ECDSA_P256 -> Curve.NIST256P1
            SigningAlgorithm.ECDSA_secp256k1 -> Curve.SECP256K1
        }
    }

    override val key: KeyPair
        get() = throw UnsupportedOperationException("Use publicKey() and privateKey() methods instead")

    override val secret: ByteArray = mnemonicString.toByteArray()
    override val advance: Any = Unit
    override val keyType: KeyType = KeyType.SEED_PHRASE
    override val isHardwareBacked: Boolean = false

    override val id: String
        get() = keyPair?.public?.encoded?.let {
            BaseEncoding.base64().encode(it)
        } ?: java.util.UUID.randomUUID().toString()

    override val mnemonic: List<String> = mnemonicString.split(" ")

    override fun deriveKey(index: Int): KeyProtocol {
        val newPath = derivationPath.replaceAfterLast("/", index.toString())
        val derivedKeyPair = deriveKeyPair(newPath)
        val twPrivateKey = TWPrivateKey(derivedKeyPair.private.encoded)
        return PrivateKey(twPrivateKey, storage)
    }

    private fun deriveKeyPair(path: String, signAlgo: SigningAlgorithm = SigningAlgorithm.ECDSA_secp256k1): KeyPair {
        var derivedPrivateKey: wallet.core.jni.PrivateKey?
        var publicKey: wallet.core.jni.PublicKey?
        try {
            val curve = getCurveForAlgorithm(signAlgo)
            derivedPrivateKey = hdWallet.getKeyByCurve(curve, path)
            publicKey = when (signAlgo) {
                SigningAlgorithm.ECDSA_P256 -> derivedPrivateKey.getPublicKeyNist256p1().uncompressed()
                SigningAlgorithm.ECDSA_secp256k1 -> derivedPrivateKey.getPublicKeySecp256k1(false)
            }

            val keyFactory = KeyFactory.getInstance("EC")
            val privateKeySpec = PKCS8EncodedKeySpec(derivedPrivateKey.data())
            val publicKeySpec = X509EncodedKeySpec(publicKey.data())

            return KeyPair(
                keyFactory.generatePublic(publicKeySpec),
                keyFactory.generatePrivate(privateKeySpec)
            )
        } catch (e: Exception) {
            Log.e(TAG, "Key derivation failed", e)
            throw WalletError.InitPrivateKeyFailed
        } finally {
            // Secure cleanup
            derivedPrivateKey = null
            publicKey = null
        }
    }

    override suspend fun create(advance: Any, storage: StorageProtocol): KeyProtocol {
        NativeLibraryManager.throwIfNotLoaded()
        val mnemonic = BIP39.generate()
        val keyPair = deriveKeyPair(DEFAULT_DERIVATION_PATH)
        return SeedPhraseKey(mnemonic, "", DEFAULT_DERIVATION_PATH, keyPair, storage)
    }

    override suspend fun create(storage: StorageProtocol): KeyProtocol {
        return create(Unit, storage)
    }

    override suspend fun createAndStore(id: String, password: String, storage: StorageProtocol): KeyProtocol {
        val key = create(storage)
        key.store(id, password)
        return key
    }

    override suspend fun get(id: String, password: String, storage: StorageProtocol): KeyProtocol {
        val encryptedData = storage.get(id) ?: throw WalletError.EmptyKeychain
        val cipher = ChaChaPolyCipher(password)
        val keyDataStr = String(cipher.decrypt(encryptedData), Charsets.UTF_8)
        val keyData = Json.decodeFromString<KeyData>(keyDataStr)
        val keyPair = deriveKeyPair(keyData.path)
        return SeedPhraseKey(keyData.mnemonic, keyData.passphrase, keyData.path, keyPair, storage, keyData.length)
    }

    override suspend fun restore(secret: ByteArray, storage: StorageProtocol): KeyProtocol {
        val mnemonic = String(secret, Charsets.UTF_8)
        if (!BIP39.isValid(mnemonic)) {
            throw WalletError.InvalidMnemonic
        }
        val keyPair = deriveKeyPair(DEFAULT_DERIVATION_PATH)
        return SeedPhraseKey(mnemonic, "", DEFAULT_DERIVATION_PATH, keyPair, storage)
    }

    override fun publicKey(signAlgo: SigningAlgorithm): ByteArray? {
        NativeLibraryManager.throwIfNotLoaded()
        
        var twPriv: wallet.core.jni.PrivateKey? = null
        var pubKey: wallet.core.jni.PublicKey? = null
        try {
            val curve = getCurveForAlgorithm(signAlgo)
            twPriv = hdWallet.getKeyByCurve(curve, derivationPath)
            pubKey = when (signAlgo) {
                SigningAlgorithm.ECDSA_P256 -> twPriv.getPublicKeyNist256p1().uncompressed()
                SigningAlgorithm.ECDSA_secp256k1 -> twPriv.getPublicKeySecp256k1(false)
                else -> null
            }
            return pubKey?.data()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get public key", e)
            return null
        } finally {
            // Secure cleanup - help prevent memory leaks
            try {
                twPriv?.let { 
                    // Clear any cached private key data
                    System.gc()
                }
                pubKey?.let {
                    // Clear any cached public key data  
                    System.gc()
                }
            } catch (e: Exception) {
                Log.w(TAG, "Cleanup warning: ${e.message}")
            }
        }
    }

    override fun privateKey(signAlgo: SigningAlgorithm): ByteArray? {
        NativeLibraryManager.throwIfNotLoaded()
        
        var twPriv: wallet.core.jni.PrivateKey? = null
        try {
            val curve = getCurveForAlgorithm(signAlgo)
            twPriv = hdWallet.getKeyByCurve(curve, derivationPath)
            return twPriv.data()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get private key", e)
            return null
        } finally {
            // Secure cleanup
            try {
                twPriv?.let {
                    // Force garbage collection to clear sensitive data
                    System.gc()
                }
            } catch (e: Exception) {
                Log.w(TAG, "Cleanup warning: ${e.message}")
            }
        }
    }

    override suspend fun sign(data: ByteArray, signAlgo: SigningAlgorithm, hashAlgo: HashingAlgorithm): ByteArray {
        if (keyPair == null) throw WalletError.EmptySignKey
        
        var twPriv: wallet.core.jni.PrivateKey?
        try {
            val curve = getCurveForAlgorithm(signAlgo)
            twPriv = hdWallet.getKeyByCurve(curve, derivationPath)
            val hashed = HasherImpl.hash(data, hashAlgo)
            val fullSignature = twPriv.sign(hashed, curve)

            // 1) Drop recovery byte if present
            val sig = if (signAlgo == SigningAlgorithm.ECDSA_secp256k1 && fullSignature.size == 65) {
                fullSignature.copyOfRange(0, 64)
            } else {
                fullSignature
            }

            return sig
        } catch (e: Exception) {
            Log.e(TAG, "Signing failed", e)
            throw WalletError.SignError
        } finally {
            // Secure cleanup
            twPriv = null
        }
    }

    override fun isValidSignature(signature: ByteArray, message: ByteArray, signAlgo: SigningAlgorithm, hashAlgo: HashingAlgorithm): Boolean {
        if (keyPair == null) return false

        var twPriv: wallet.core.jni.PrivateKey?
        var pubKey: wallet.core.jni.PublicKey?
        try {
            val curve = getCurveForAlgorithm(signAlgo)
            twPriv = hdWallet.getKeyByCurve(curve, derivationPath)
            if (signAlgo != SigningAlgorithm.ECDSA_P256 && signAlgo != SigningAlgorithm.ECDSA_secp256k1) return false
            pubKey = when (signAlgo) {
                SigningAlgorithm.ECDSA_P256 -> twPriv.getPublicKeyNist256p1().uncompressed()
                SigningAlgorithm.ECDSA_secp256k1 -> twPriv.getPublicKeySecp256k1(false)
            }
            val hashed = HasherImpl.hash(message, hashAlgo)
            return pubKey.verify(hashed, signature)
        } catch (e: Exception) {
            Log.e(TAG, "Signature verification failed", e)
            return false
        } finally {
            // Secure cleanup
            twPriv = null
            pubKey = null
        }
    }

    override suspend fun store(id: String, password: String) {
        val keyData = createKeyData()
        val encryptedData = try {
            encryptData(keyData, password)
        } catch (e: Exception) {
            Log.e(TAG, "Encryption failed", e)
            throw WalletError.InitChaChaPolyFailed
        }
        storage.set(id, encryptedData)
    }

    override suspend fun remove(id: String) {
        storage.remove(id)
    }

    override fun allKeys(): List<String> {
        return storage.allKeys
    }

    private fun createKeyData(): ByteArray {
        val data = mapOf(
            "mnemonic" to mnemonicString,
            "passphrase" to passphrase,
            "path" to derivationPath,
            "length" to seedPhraseLength
        )
        return Json.encodeToString(data).toByteArray()
    }

    private fun encryptData(data: ByteArray, password: String): ByteArray {
        val cipher = ChaChaPolyCipher(password)
        return cipher.encrypt(data)
    }

    override fun ethAddress(index: Int): String = withEthereumPrivateKey(index) {
        CoinType.ETHEREUM.deriveAddress(it)
    }

    override fun ethPublicKey(index: Int): ByteArray = withEthereumPrivateKey(index) {
        it.getPublicKeySecp256k1(false).data()
    }

    override fun ethPrivateKey(index: Int): ByteArray = withEthereumPrivateKey(index) {
        it.data()
    }

    override fun ethSignDigest(digest: ByteArray, index: Int): ByteArray = withEthereumPrivateKey(index) {
        EthereumSignatureUtils.validateDigest(digest)
        try {
            val signature = it.sign(digest, Curve.SECP256K1)
            EthereumSignatureUtils.normalize(signature)
        } catch (e: Exception) {
            Log.e(TAG, "Ethereum signing failed", e)
            throw WalletError.SignError
        }
    }

    private inline fun <T> withEthereumPrivateKey(index: Int, block: (TWPrivateKey) -> T): T {
        NativeLibraryManager.throwIfNotLoaded()
        if (index < 0) throw WalletError.UnsupportedEthereumDerivation
        var twPriv: TWPrivateKey? = null
        return try {
            val path = ethereumDerivationPath(index)
            twPriv = hdWallet.getKeyByCurve(Curve.SECP256K1, path)
            block(twPriv!!)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to derive Ethereum key", e)
            throw WalletError.InvalidEthereumDerivationPath
        } finally {
            twPriv = null
        }
    }

    private fun ethereumDerivationPath(index: Int): String {
        if (index < 0) throw WalletError.UnsupportedEthereumDerivation
        return "$ETH_DERIVATION_PREFIX/$index"
    }
}

// Helper class for returning four values
private data class Quadruple<A, B, C, D>(
    val first: A,
    val second: B,
    val third: C,
    val fourth: D
) 

File: iOS/FlowWalletKit/Sources/Keys/SeedPhraseKey.swift
/// FlowWalletKit - BIP39 Seed Phrase Key Implementation
///
/// This module implements key management using BIP39 seed phrases.
/// It provides functionality for:
/// - Creating and managing HD wallets
/// - Deriving keys using BIP44 paths
/// - Secure storage of seed phrases
/// - Transaction signing with derived keys

import CryptoKit
import Flow
import Foundation
import KeychainAccess
import WalletCore

/// Implementation of KeyProtocol using BIP39 seed phrases
public class SeedPhraseKey: KeyProtocol {
    // MARK: - Types
    
    /// Advanced options for seed phrase key creation
    public struct AdvanceOption {
        /// BIP44 derivation path
        let derivationPath: String
        /// Length of the seed phrase (12, 15, 18, 21, 24 words)
        let seedPhraseLength: BIP39.SeedPhraseLength
        /// Optional passphrase for additional security
        let passphrase: String
    }

    /// Storage model for seed phrase data
    public struct KeyData: Codable {
        /// BIP39 mnemonic words
        let mnemonic: String
        /// BIP44 derivation path
        let derivationPath: String
        /// Optional passphrase
        let passphrase: String
    }

    // MARK: - Properties

    /// Storage implementation for persisting key data
    public var storage: any StorageProtocol
    /// Type identifier for this key implementation
    public var keyType: KeyType = .seedPhrase
    /// BIP44 derivation path (default: Flow path m/44'/539'/0'/0/0)
    static public var derivationPath = "m/44'/539'/0'/0/0"
    /// Optional passphrase for additional security
    static public var passphrase: String = ""
    
    /// BIP44 derivation path (default: Flow path m/44'/539'/0'/0/0)
    public var derivationPath: String
    /// Optional passphrase for additional security
    public var passphrase: String
    
    /// Default base derivation path for Ethereum accounts
    public static var ethBaseDerivationPath = DerivationPath(purpose: .bip44,
                                                             coin: CoinType.ethereum.slip44Id,
                                                             account: 0,
                                                             change: 0,
                                                             address: 0).description

    /// Default seed phrase length (12 words)
    public static let defaultSeedPhraseLength: BIP39.SeedPhraseLength = .twelve

    /// Underlying HD wallet implementation
    public let hdWallet: HDWallet

    // MARK: - Initialization

    /// Initialize a seed phrase key
    /// - Parameters:
    ///   - hdWallet: HD wallet implementation
    ///   - storage: Storage implementation
    ///   - derivationPath: BIP44 derivation path
    ///   - passphrase: Optional passphrase
    ///   - seedPhraseLength: Length of seed phrase
    public init(hdWallet: HDWallet,
                storage: any StorageProtocol,
                derivationPath: String = SeedPhraseKey.derivationPath,
                passphrase: String = SeedPhraseKey.passphrase
    ) {
        self.hdWallet = hdWallet
        self.storage = storage
        self.derivationPath = derivationPath
        self.passphrase = passphrase
    }

    // MARK: - Key Creation

    /// Create a new key with advanced options
    /// - Parameters:
    ///   - advance: Advanced creation options
    ///   - storage: Storage implementation
    /// - Returns: New seed phrase key
    /// - Throws: FWKError if HD wallet creation fails
    public static func create(_ advance: AdvanceOption, storage: any StorageProtocol) throws -> SeedPhraseKey {
        guard let hdWallet = HDWallet(strength: advance.seedPhraseLength.strength, passphrase: advance.passphrase) else {
            throw FWKError.initHDWalletFailed
        }

        let key = SeedPhraseKey(hdWallet: hdWallet,
                                storage: storage,
                                derivationPath: advance.derivationPath,
                                passphrase: advance.passphrase)
        return key
    }
    
    /// Create a new key with advanced options
    /// - Parameters:
    ///   - seedphrase: Seed Phrase
    ///   - storage: Storage implementation
    /// - Returns: New seed phrase key
    /// - Throws: FWKError if HD wallet creation fails
    public static func create(_ seedphrase: String,
                              derivationPath: String = SeedPhraseKey.derivationPath,
                              passphrase: String = SeedPhraseKey.passphrase,
                              storage: any StorageProtocol) throws -> SeedPhraseKey {
        guard let hdWallet = HDWallet(mnemonic: seedphrase, passphrase: passphrase) else {
            throw FWKError.initHDWalletFailed
        }
        
        let key = SeedPhraseKey(hdWallet: hdWallet,
                                storage: storage,
                                derivationPath: derivationPath,
                                passphrase: passphrase)
        return key
    }

    /// Create a new key with default options
    /// - Parameter storage: Storage implementation
    /// - Returns: New seed phrase key
    /// - Throws: FWKError if HD wallet creation fails
    public static func create(storage: any StorageProtocol) throws -> SeedPhraseKey {
        guard let hdWallet = HDWallet(strength: SeedPhraseKey.defaultSeedPhraseLength.strength, passphrase: "") else {
            throw FWKError.initHDWalletFailed
        }
        return SeedPhraseKey(hdWallet: hdWallet, storage: storage)
    }

    /// Create and store a new key
    /// - Parameters:
    ///   - id: Unique identifier for the key
    ///   - password: Password for encrypting the key
    ///   - storage: Storage implementation
    /// - Returns: New seed phrase key
    /// - Throws: FWKError if creation or storage fails
    public static func createAndStore(id: String, password: String, storage: any StorageProtocol) throws -> SeedPhraseKey {
        guard let hdWallet = HDWallet(strength: SeedPhraseKey.defaultSeedPhraseLength.strength, passphrase: "") else {
            throw FWKError.initHDWalletFailed
        }
        let key = SeedPhraseKey(hdWallet: hdWallet, storage: storage)
        try key.store(id: id, password: password)
        return key
    }

    // MARK: - Key Recovery

    /// Retrieve a stored key
    /// - Parameters:
    ///   - id: Unique identifier for the key
    ///   - password: Password for decrypting the key
    ///   - storage: Storage implementation
    /// - Returns: Retrieved seed phrase key
    /// - Throws: FWKError if retrieval fails
    public static func get(id: String, password: String, storage: any StorageProtocol) throws -> SeedPhraseKey {
        guard let data = try storage.get(id) else {
            throw FWKError.emptyKeychain
        }

        guard let cipher = ChaChaPolyCipher(key: password) else {
            throw FWKError.initChaChapolyFailed
        }

        let entropyData = try cipher.decrypt(combinedData: data)
        let model = try JSONDecoder().decode(KeyData.self, from: entropyData)

        guard let hdWallet = HDWallet(mnemonic: model.mnemonic, passphrase: model.passphrase) else {
            throw FWKError.initHDWalletFailed
        }
        
        return SeedPhraseKey(hdWallet: hdWallet, storage: storage, derivationPath: model.derivationPath, passphrase: model.passphrase)
    }

    // MARK: - Cryptographic Operations

    /// Verify a signature
    /// - Parameters:
    ///   - signature: Signature to verify
    ///   - message: Original message that was signed
    ///   - signAlgo: Signature algorithm used
    /// - Returns: Whether the signature is valid
    public func isValidSignature(signature: Data, message: Data, signAlgo: Flow.SignatureAlgorithm) -> Bool {
        guard let pubK = try? getPublicKey(signAlgo: signAlgo) else {
            return false
        }
        return pubK.verify(signature: signature, message: message)
    }

    /// Store the key securely
    /// - Parameters:
    ///   - id: Unique identifier for the key
    ///   - password: Password for encrypting the key
    /// - Throws: FWKError if storage fails
    public func store(id: String, password: String) throws {
        guard let cipher = ChaChaPolyCipher(key: password) else {
            throw FWKError.initChaChapolyFailed
        }
        let model = KeyData(mnemonic: hdWallet.mnemonic,
                            derivationPath: SeedPhraseKey.derivationPath,
                            passphrase: passphrase)
        let data = try JSONEncoder().encode(model)
        let encrypted = try cipher.encrypt(data: data)
        try storage.set(id, value: encrypted)
    }

    /// Restore a key from secret material
    /// - Parameters:
    ///   - secret: Secret key data
    ///   - storage: Storage implementation
    /// - Returns: Restored seed phrase key
    /// - Throws: FWKError if restoration fails
    public static func restore(secret: KeyData, storage: any StorageProtocol) throws -> SeedPhraseKey {
        guard let wallet = HDWallet(mnemonic: secret.mnemonic, passphrase: secret.passphrase) else {
            throw FWKError.restoreWalletFailed
        }

        let key = SeedPhraseKey(hdWallet: wallet, storage: storage,
                                derivationPath: secret.derivationPath, passphrase: secret.passphrase)
        return key
    }

    /// Get the public key for a signature algorithm
    /// - Parameter signAlgo: Signature algorithm
    /// - Returns: Public key data
    public func publicKey(signAlgo: Flow.SignatureAlgorithm) -> Data? {
        guard let pubK = try? getPublicKey(signAlgo: signAlgo) else {
            return nil
        }
        return pubK.uncompressed.data.dropFirst() // 04
    }
    
    /// Get the private key for a signature algorithm
    /// - Parameter signAlgo: Signature algorithm
    /// - Returns: Private key data
    public func privateKey(signAlgo: Flow.SignatureAlgorithm) -> Data? {
        guard let curve = signAlgo.WCCurve else {
            return nil
        }
        var pk = hdWallet.getKeyByCurve(curve: curve, derivationPath: SeedPhraseKey.derivationPath)
        defer { pk = WalletCore.PrivateKey() }
        return pk.data
    }

    /// Sign data using specified algorithms
    /// - Parameters:
    ///   - data: Data to sign
    ///   - signAlgo: Signature algorithm
    ///   - hashAlgo: Hash algorithm
    /// - Returns: Signature data
    /// - Throws: FWKError if signing fails
    public func sign(data: Data, signAlgo: Flow.SignatureAlgorithm, hashAlgo: Flow.HashAlgorithm) throws -> Data {
        let hashed = try hashAlgo.hash(data: data)

        guard let curve = signAlgo.WCCurve else {
            throw FWKError.unsupportSignatureAlgorithm
        }

        var pk = hdWallet.getKeyByCurve(curve: curve, derivationPath: SeedPhraseKey.derivationPath)
        defer { pk = WalletCore.PrivateKey() }
        guard let signature = pk.sign(digest: hashed, curve: curve) else {
            throw FWKError.signError
        }
        return signature.dropLast()
    }

    // MARK: - Private Methods

    /// Get the public key implementation for a signature algorithm
    /// - Parameter signAlgo: Signature algorithm
    /// - Returns: Public key implementation
    /// - Throws: FWKError if algorithm is not supported
    private func getPublicKey(signAlgo: Flow.SignatureAlgorithm) throws -> PublicKey {
        guard let pubKey = hdWallet.getPublicKey(signAlgo: signAlgo) else {
            throw FWKError.unsupportSignatureAlgorithm
        }
        return pubKey
    }
}

// MARK: - Ethereum Support

extension SeedPhraseKey: EthereumKeyProtocol {
    
    public func ethAddress(index: UInt32) throws -> String {
        try withEthereumPrivateKey(index: index) { key in
            CoinType.ethereum.deriveAddress(privateKey: key)
        }
    }
    
    public func ethPublicKey(index: UInt32) throws -> Data {
        try withEthereumPrivateKey(index: index) { key in
            key.getPublicKeySecp256k1(compressed: false).uncompressed.data
        }
    }
    
    public func ethPrivateKey(index: UInt32) throws -> Data {
        try withEthereumPrivateKey(index: index) { key in
            key.data
        }
    }
    
    public func ethSign(digest: Data, index: UInt32) throws -> Data {
        try withEthereumPrivateKey(index: index) { key in
            try validateEthereumDigest(digest)
            guard let signature = key.sign(digest: digest, curve: .secp256k1) else {
                throw FWKError.signError
            }
            // Signature is 65 bytes: [r(32) | s(32) | v(1)]
            return try normalizeEthereumSignature(signature)
        }
    }
    
    private func ethereumDerivationPath(index: UInt32) throws -> DerivationPath {
        guard let basePath = DerivationPath(SeedPhraseKey.ethBaseDerivationPath) else {
            throw FWKError.invalidEthereumDerivationPath
        }
        return DerivationPath(purpose: basePath.purpose,
                              coin: basePath.coin,
                              account: basePath.account,
                              change: basePath.change,
                              address: index)
    }
    
    private func withEthereumPrivateKey<T>(index: UInt32,
                                           _ body: (WalletCore.PrivateKey) throws -> T) throws -> T {
        let path = try ethereumDerivationPath(index: index)
        var key = hdWallet.getKey(coin: .ethereum, derivationPath: path.description)
        defer { key = WalletCore.PrivateKey() }
        return try body(key)
    }
}
